### PR TITLE
update cache4statscache

### PR DESCRIPTION
**Pull request by miamia0**
update badger cache used for statistic cache

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -62,7 +62,8 @@ type Config struct {
 	// Cost evaluates a value and outputs a corresponding cost. This function
 	// is ran after Set is called for a new item or an item update with a cost
 	// param of 0.
-	Cost func(value interface{}) int64
+	OnInsert func(key uint64, cost int64)
+	Cost     func(value interface{}) int64
```
Comment (bobotu) [inline cache/cache.go:66]:
```suggestion
	Cost     func(value interface{}) int64
	OnInsert func(key uint64, cost int64)
```

Swap lines to match comments

File: cache/policy.go [hunk 2]
```diff
@@ -248,7 +260,7 @@ func (p *sampledLFU) del(key uint64) {
 }
 
 func (p *sampledLFU) add(key uint64, cost int64) {
-	p.keyCosts[key] = cost
+	p.keyCosts[key] += cost
```
Comment (coocood) [inline cache/policy.go:263]: What if the same key being added multiple times?

File: cache/policy.go [hunk 3]
```diff
@@ -85,32 +85,45 @@ func (p *policy) Push(keys []uint64) bool {
 		return false
 	}
 }
-
-func (p *policy) Add(key uint64, cost int64) ([]*item, bool) {
+func (p *sampledLFU) checkUpdateIfHas(key uint64, cost int64) (truecost int64, has bool) {
```
Comment (bobotu) [inline cache/policy.go:88]:
```suggestion

func (p *sampledLFU) checkUpdateIfHas(key uint64, cost int64) (truecost int64, has bool) {
```

File: cache/policy.go [hunk 4]
```diff
@@ -85,32 +85,45 @@ func (p *policy) Push(keys []uint64) bool {
...
+	if prev, found := p.keyCosts[key]; found {
+		// update the cost of an existing key, but don't worry about evicting,
+		// evictions will be handled the next time a new item is added
+		return cost - prev, true
+	}
+	return cost, false
+}
```
Comment (bobotu) [inline cache/policy.go:95]:
```suggestion
}

```

File: cache/policy.go [hunk 5]
```diff
@@ -128,15 +141,10 @@ func (p *policy) Add(key uint64, cost int64) ([]*item, bool) {
 		minKey, minHits, minId := uint64(0), int64(math.MaxInt64), 0
 		for i, pair := range sample {
 			// look up hit count for sample key
-			if hits := p.admit.Estimate(pair.key); hits < minHits {
+			if hits := p.admit.Estimate(pair.key); hits < minHits && pair.key != key {
 				minKey, minHits, minId = pair.key, hits, i
 			}
 		}
-		// if the incoming item isn't worth keeping in the policy, reject.
```
Comment (bobotu) [inline cache/policy.go:135]: The tinyLFU admission logical is removed?

File: cache/policy.go (hunk 2 above)
Comment (coocood) [reply cache/policy.go:263]: Seems `add` is never called when key already exists.

**Comment by coocood** [comment]
@miamia0 
Why do we need OnInsert and the true cost?

**Comment by miamia0** [comment]
> @miamia0
> Why do we need OnInsert and the true cost?

i want to detect cost change for upper memory Tracker. so want to add a interface with insert.

File: cache/policy.go (hunk 2 above)
Comment (miamia0) [reply cache/policy.go:263]: when key already exists, cost would be the cost change of value.

File: cache/cache.go [hunk 6]
```diff
@@ -56,6 +56,8 @@ type Config struct {
 	// only set this flag to true when testing or throughput performance isn't a
 	// major factor.
 	Metrics bool
+	// NotEvictNew determines if new k-v can be evict when Set
+	NotEvictNew bool
```
Comment (coocood) [inline cache/cache.go:60]: `EvictNew` is easier to read.

File: cache/policy.go [hunk 7]
```diff
@@ -94,19 +110,27 @@ func (p *policy) Add(key uint64, cost int64) ([]*item, bool) {
 		return nil, false
 	}
 	// we don't need to go any further if the item is already in the cache
-	if has := p.evict.updateIfHas(key, cost); has {
-		return nil, true
-	}
+	cost, has := p.evict.checkUpdateIfHas(key, cost)
```
Comment (coocood) [inline cache/policy.go:113]: Can we revert this change after `GetMemoryUsed` is added?

File: cache/policy.go [hunk 8]
```diff
@@ -86,6 +88,20 @@ func (p *policy) Push(keys []uint64) bool {
 	}
 }
 
+// GetMemoryUsed return the used memory.
+func (p *policy) GetMemoryUsed() int64 {
```
Comment (coocood) [inline cache/policy.go:92]: GetTotalCost is better.

File: cache/cache.go (hunk 6 above)
Comment (bobotu) [reply cache/cache.go:60]: We need to introduce something like `DefaultConfig` to enable `EvictNew`.

**Comment by miamia0** [comment]
change implement, no need to change cache anymore: ) thanks.
